crates/sui-source-validation/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+    fmt::{Debug, Display},
+    str::FromStr,
+};
+
+use move_compiler::compiled_unit::CompiledUnitEnum;
+use move_core_types::account_address::AccountAddress;
+use move_package::compilation::compiled_package::CompiledPackage;
+use move_symbol_pool::Symbol;
+
+use sui_sdk::{
+    rpc_types::{SuiRawData, SuiRawMoveObject, SuiRawMovePackage},
+    ReadApi,
+};
+use sui_types::{
+    base_types::{ObjectID, ObjectIDParseError},
+    error::SuiError,
+};
+
+#[derive(Clone, Debug)]
+pub struct DependencyVerificationResult {
+    pub verified_dependencies: HashMap<AccountAddress, Dependency>,
+}
+
+#[derive(Debug)]
+pub enum DependencyVerificationError {
+    /// Could not convert a dependencies' resolved Sui address to a Sui object ID
+    ObjectIdFromAddressFailure(ObjectIDParseError),
+    /// Could not read a dependencies' on-chain object
+    DependencyObjectReadFailure(anyhow::Error),
+    /// Dependency object does not exist or was deleted
+    SuiObjectRefFailure(SuiError),
+    /// Dependency address contains a Sui object, not a Move package
+    ObjectFoundWhenPackageExpected(ObjectID, SuiRawMoveObject),
+    /// A local dependency was not found
+    ///
+    /// params:  package, module
+    LocalDependencyNotFound(Symbol, Option<Symbol>),
+    /// An on-chain version of a local dependency was not found
+    ///
+    /// params:  package, module
+    OnChainDependencyNotFound(Symbol, Option<Symbol>),
+    /// Local dependencies have a different number of modules than on-chain
+    ///
+    /// params:  expected count, on-chain count, missing
+    ModuleCountMismatch(usize, usize, Vec<String>),
+    /// A local dependency module did not match its on-chain version
+    ///
+    /// params:  package, module, address
+    ModuleBytecodeMismatch(String, String, AccountAddress),
+}
+
+impl Display for DependencyVerificationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self, f)
+    }
+}
+
+#[derive(Debug)]
+pub struct BytecodeSourceVerifier<'a> {
+    pub verbose: bool,
+    rpc_client: &'a ReadApi,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Dependency {
+    pub symbol: String,
+    pub module_bytes: BTreeMap<String, Vec<u8>>,
+}
+
+type ModuleBytesMap = HashMap<Symbol, (AccountAddress, HashMap<Symbol, Vec<u8>>)>;
+
+impl<'a> BytecodeSourceVerifier<'a> {
+    pub fn new(rpc_client: &'a ReadApi, verbose: bool) -> Self {
+        BytecodeSourceVerifier {
+            verbose,
+            rpc_client,
+        }
+    }
+
+    /// Verify that all local Move package dependencies' bytecode matches
+    /// the bytecode at the address specified on the Sui network we are publishing to.
+    pub async fn verify_deployed_dependencies(
+        &mut self,
+        compiled_package: CompiledPackage,
+    ) -> Result<DependencyVerificationResult, DependencyVerificationError> {
+        let compiled_dep_map = Self::get_module_bytes_map(&compiled_package);
```

Comment:
If this is a "static" function, can it be moved out of the `impl` block (i.e. be made a top-level non-public function?)
```suggestion
        let compiled_dep_map = get_module_bytes_map(&compiled_package);
```